File: apps/execution/evaluator.py
"""
Deterministic, semantically matching Python AST evaluator for execution-side checks.
Supports comparison/null semantics, arithmetic, relative paths, and indexed-repeat consistently.
"""

from typing import Any, Dict, Optional


def resolve_relative_path(path: str) -> str:
    """
    Translates relative path references using standard rules.
    """
    if not isinstance(path, str):
        return path
    if path.startswith("../"):
        return "/clinical_data/subject/" + path[3:]
    return path


def get_node_attribute(node: Any, attr: str) -> Any:
    """
    Safely retrieves an attribute or dictionary key from an AST node.
    Supports both Pydantic models (or class objects) and raw dictionaries.
    """
    if node is None:
        return None
    if isinstance(node, dict):
        return node.get(attr, None)
    return getattr(node, attr, None)


def evaluate_ast(
    node: Any,
    context: Optional[Dict[str, Any]] = None,
    current_indices: Optional[Dict[str, int]] = None,
) -> Any:
    """
    Deterministically evaluates an ExpressionNode AST against the given data context.
    Matches the client-side JavaScript evaluator semantics exactly.
    """
    if node is None:
        return None
    if context is None:
        context = {}
    if current_indices is None:
        current_indices = {}

    # Extract attributes supporting both dict and object structures
    node_type = (
        get_node_attribute(node, "node_type") or get_node_attribute(node, "type") or ""
    )
    value = get_node_attribute(node, "value")
    operands = (
        get_node_attribute(node, "children")
        or get_node_attribute(node, "operands")
        or []
    )
    field_ref = get_node_attribute(node, "field_ref")

    # 1. LITERAL / CONSTANT
    if node_type in ("LITERAL", "constant"):
        return value

    # 2. XPATH / FIELD_REF
    if node_type in ("XPATH", "field_ref"):
        path = ""
        if field_ref is not None:
            path = get_node_attribute(field_ref, "field_id") or ""
        elif isinstance(value, str):
            path = value

        if not path:
            return None

        # Direct context lookup
        if path in context:
            return context[path]

        # Resolved path lookup
        resolved_path = resolve_relative_path(path)
        if resolved_path in context:
            return context[resolved_path]

        # Bare field name lookup fallback (e.g. /clinical_data/vssbp -> vssbp)
        bare_field = path.split("/")[-1]
        if bare_field in context:
            return context[bare_field]

        return None

    # 3. OPERATOR / LOGICAL / COMPARISON
    if node_type in ("OPERATOR", "logical", "comparison"):
        operator_val = get_node_attribute(node, "operator") or value
        if not isinstance(operator_val, str):
            return None
        operator = operator_val.lower()

        # Logical operators (Kleene 3-valued logic)
        if operator == "not":
            if not operands:
                return None
            child_val = evaluate_ast(operands[0], context, current_indices)
            if child_val is None:
                return None
            return not child_val

        if operator == "and":
            # Kleene logic: if any operand is False, return False.
            # Else if any is None, return None. Else return True.
            has_none = False
            for child in operands:
                child_val = evaluate_ast(child, context, current_indices)
                if child_val is False:
                    return False
                if child_val is None:
                    has_none = True
            return None if has_none else True

        if operator == "or":
            # Kleene logic: if any operand is True, return True.
            # Else if any is None, return None. Else return False.
            has_none = False
            for child in operands:
                child_val = evaluate_ast(child, context, current_indices)
                if child_val is True:
                    return True
                if child_val is None:
                    has_none = True
            return None if has_none else False

        # Arithmetic and Comparison requires at least 2 operands
        if len(operands) < 2:
            return None

        left_val = evaluate_ast(operands[0], context, current_indices)
        right_val = evaluate_ast(operands[1], context, current_indices)

        # Arithmetic Operators (with null-safety)
        if operator in ("+", "-", "*", "/"):
            if left_val is None or right_val is None:
                return None
            try:
                l_num = float(left_val)
                r_num = float(right_val)
            except (ValueError, TypeError):
                return None

            if operator == "+":
                return l_num + r_num
            elif operator == "-":
                return l_num - r_num
            elif operator == "*":
                return l_num * r_num
            elif operator == "/":
                if r_num == 0.0:
                    return None  # Safe division by zero
                return l_num / r_num

        # Comparison Operators (with null-safety)
        if operator in ("==", "!=", "<", "<=", ">", ">="):
            if operator == "==":
                return left_val == right_val
            if operator == "!=":
                return left_val != right_val

            # Ordered comparison with None always returns False
            if left_val is None or right_val is None:
                return False

            try:
                l_num = float(left_val)
                r_num = float(right_val)
                use_numeric = True
            except (ValueError, TypeError):
                use_numeric = False

            l_cmp = l_num if use_numeric else str(left_val)
            r_cmp = r_num if use_numeric else str(right_val)

            if operator == "<":
                return l_cmp < r_cmp
            elif operator == "<=":
                return l_cmp <= r_cmp
            elif operator == ">":
                return l_cmp > r_cmp
            elif operator == ">=":
                return l_cmp >= r_cmp

        return None

    # 4. FUNCTION
    if node_type in ("FUNCTION", "function"):
        func_val = get_node_attribute(node, "operator") or value
        if not isinstance(func_val, str):
            return None
        func_name = func_val.lower()

        if func_name in ("is_empty", "empty"):
            if len(operands) != 1:
                return None
            child_val = evaluate_ast(operands[0], context, current_indices)
            if child_val is None:
                return True
            if isinstance(child_val, str) and child_val.strip() == "":
                return True
            return False

        if func_name == "is_not_empty":
            if len(operands) != 1:
                return None
            child_val = evaluate_ast(operands[0], context, current_indices)
            if child_val is None:
                return False
            if isinstance(child_val, str) and child_val.strip() == "":
                return False
            return True

        if func_name == "indexed-repeat":
            if len(operands) != 3:
                return None
            target_field_node = operands[0]
            repeat_group_node = operands[1]
            index_node = operands[2]

            target_path = (
                get_node_attribute(target_field_node, "value")
                or (
                    get_node_attribute(target_field_node, "field_ref")
                    and get_node_attribute(
                        get_node_attribute(target_field_node, "field_ref"), "field_id"
                    )
                )
                or ""
            )
            repeat_group = (
                get_node_attribute(repeat_group_node, "value")
                or (
                    get_node_attribute(repeat_group_node, "field_ref")
                    and get_node_attribute(
                        get_node_attribute(repeat_group_node, "field_ref"), "field_id"
                    )
                )
                or ""
            )
            index_val = evaluate_ast(index_node, context, current_indices)

            try:
                target_index = int(index_val)
            except (ValueError, TypeError):
                return None

            field_name = target_path.split("/")[-1]
            indexed_path = f"{repeat_group}[{target_index}]/{field_name}"

            return context.get(indexed_path, None)

        return None

    return None

File: apps/web/src/evaluator.js
/**
 * Client-side AST evaluator for clinical rules (relevant skip logic and constraints).
 */
import { validateField as uiValidateField } from "ui";

class SimpleLRUCache {
  constructor(capacity = 200) {
    this.capacity = capacity;
    this.cache = new Map();
  }

  get(key) {
    if (!this.cache.has(key)) return undefined;
    const value = this.cache.get(key);
    this.cache.delete(key);
    this.cache.set(key, value);
    return value;
  }

  put(key, value) {
    if (this.cache.has(key)) {
      this.cache.delete(key);
    } else if (this.cache.size >= this.capacity) {
      const firstKey = this.cache.keys().next().value;
      this.cache.delete(firstKey);
    }
    this.cache.set(key, value);
  }

  has(key) {
    return this.cache.has(key);
  }

  resize(newCapacity) {
    this.capacity = Math.max(0, newCapacity);
    while (this.cache.size > this.capacity) {
      const firstKey = this.cache.keys().next().value;
      this.cache.delete(firstKey);
    }
  }
}

// Global initialization of compiler Cache
export const compilerCache = new SimpleLRUCache(200);

export function resizeCompilerCache(capacity) {
  compilerCache.resize(capacity);
}

export class DiagnosticRegistry {
  constructor() {
    this.listeners = new Set();
  }

  subscribe(callback) {
    this.listeners.add(callback);
    return () => {
      this.listeners.delete(callback);
    };
  }

  unsubscribe(callback) {
    this.listeners.delete(callback);
  }

  register(callback) {
    this.listeners.add(callback);
  }

  unregister(callback) {
    this.listeners.delete(callback);
  }

  emit(event) {
    for (const listener of this.listeners) {
      try {
        listener(event);
      } catch (err) {
        // Errors thrown within third-party dynamic listener callbacks must be caught internally to prevent app-level crashes or validation bypasses
        console.error("Error in diagnostic listener callback:", err);
      }
    }
  }
}

export const diagnosticRegistry = new DiagnosticRegistry();

export function registerDiagnosticListener(callback) {
  return diagnosticRegistry.subscribe(callback);
}

export function unregisterDiagnosticListener(callback) {
  diagnosticRegistry.unsubscribe(callback);
}

/**
 * Resolves a potentially relative path.
 *
 * @param {string} path
 * @returns {string} Absolute representation
 */
export function resolveRelativePath(path) {
  if (typeof path !== "string") return path;
  if (path.startsWith("../")) {
    return "/clinical_data/subject/" + path.substring(3);
  }
  return path;
}

/**
 * Deterministically evaluates an ExpressionNode AST against the given data context.
 *
 * Supports both:
 * Style A (Pydantic / Designer rules): type, operator, operands, value, field_ref
 * Style B (TDD / XForms): node_type, value, children
 *
 * @param {Object} node - AST ExpressionNode
 * @param {Object} context - Flat or nested data object (field_id -> value)
 * @param {Object} currentIndices - Map of repeat group indices
 * @returns {*} Evaluation result
 */
export function evaluateAST(node, context = {}, currentIndices = {}) {
  if (!node) return null;

  // Normalize fields between Style A and Style B
  const type = node.node_type || node.type || "";
  const valAttr = node.value;
  const rawChildren = node.children || node.operands || [];
  const fieldRef = node.field_ref;

  // 1. LITERAL / CONSTANT
  if (type === "LITERAL" || type === "constant") {
    return valAttr !== undefined ? valAttr : null;
  }

  // 2. XPATH / FIELD_REF
  if (type === "XPATH" || type === "field_ref") {
    let path = "";
    if (fieldRef && fieldRef.field_id) {
      path = fieldRef.field_id;
    } else if (typeof valAttr === "string") {
      path = valAttr;
    }

    if (!path) return null;

    // Direct lookup
    if (context[path] !== undefined) {
      return context[path];
    }

    // Relative path translation
    const resolvedPath = resolveRelativePath(path);
    if (context[resolvedPath] !== undefined) {
      return context[resolvedPath];
    }

    // Extract bare field name from path if needed (e.g. /clinical_data/.../vssbp -> vssbp)
    const bareField = path.split("/").pop();
    if (context[bareField] !== undefined) {
      return context[bareField];
    }

    return null;
  }

  // 3. OPERATOR / LOGICAL / COMPARISON
  if (type === "OPERATOR" || type === "logical" || type === "comparison") {
    const operator = (
      typeof valAttr === "string" ? valAttr : node.operator || ""
    ).toLowerCase();

    // Logical operators
    if (operator === "not") {
      if (rawChildren.length !== 1) {
        diagnosticRegistry.emit({
          type: "arity_mismatch",
          nodeType: type,
          name: operator,
          expected: 1,
          actual: rawChildren.length,
          message: `Operator '${operator}' expects exactly 1 operand, but got ${rawChildren.length}.`,
          node
        });
        return null;
      }
      const childVal = evaluateAST(rawChildren[0], context, currentIndices);
      return !childVal;
    }

    if (operator === "and") {
      // Kleene 3-valued logic: if any operand is false, return false.
      // Otherwise if any is null, return null. Else return true.
      if (rawChildren.length === 0) {
        diagnosticRegistry.emit({
          type: "arity_mismatch",
          nodeType: type,
          name: operator,
          expected: ">= 1",
          actual: 0,
          message: `Operator '${operator}' expects at least 1 operand, but got 0.`,
          node
        });
        return null;
      }
      let hasNull = false;
      for (const child of rawChildren) {
        const childVal = evaluateAST(child, context, currentIndices);
        if (childVal === false) return false;
        if (childVal === null || childVal === undefined) {
          hasNull = true;
        }
      }
      return hasNull ? null : true;
    }

    if (operator === "or") {
      // Kleene 3-valued logic: if any operand is true, return true.
      // Otherwise if any is null, return null. Else return false.
      if (rawChildren.length === 0) {
        diagnosticRegistry.emit({
          type: "arity_mismatch",
          nodeType: type,
          name: operator,
          expected: ">= 1",
          actual: 0,
          message: `Operator '${operator}' expects at least 1 operand, but got 0.`,
          node
        });
        return null;
      }
      let hasNull = false;
      for (const child of rawChildren) {
        const childVal = evaluateAST(child, context, currentIndices);
        if (childVal === true) return true;
        if (childVal === null || childVal === undefined) {
          hasNull = true;
        }
      }
      return hasNull ? null : false;
    }

    // Arithmetic and Comparison operators
    if (["+", "-", "*", "/"].includes(operator)) {
      if (rawChildren.length !== 2) {
        diagnosticRegistry.emit({
          type: "arity_mismatch",
          nodeType: type,
          name: operator,
          expected: 2,
          actual: rawChildren.length,
          message: `Operator '${operator}' expects exactly 2 operands, but got ${rawChildren.length}.`,
          node
        });
        return null;
      }
      const leftVal = evaluateAST(rawChildren[0], context, currentIndices);
      const rightVal = evaluateAST(rawChildren[1], context, currentIndices);

      // Null safety for arithmetic
      if (
        leftVal === null ||
        leftVal === undefined ||
        rightVal === null ||
        rightVal === undefined
      ) {
        return null;
      }
      const lNum = parseFloat(leftVal);
      const rNum = parseFloat(rightVal);
      if (isNaN(lNum) || isNaN(rNum)) {
        diagnosticRegistry.emit({
          type: "type_mismatch",
          nodeType: type,
          name: operator,
          expected: "numeric",
          actual: isNaN(lNum) ? (leftVal === null ? "null" : typeof leftVal) : (rightVal === null ? "null" : typeof rightVal),
          message: `Arithmetic operator '${operator}' expects numeric operands, but got non-numeric value.`,
          node
        });
        return null;
      }

      if (operator === "+") return lNum + rNum;
      if (operator === "-") return lNum - rNum;
      if (operator === "*") return lNum * rNum;
      if (operator === "/") {
        if (rNum === 0) return null; // Safe division by zero
        return lNum / rNum;
      }
    }

    // Comparison Operators (null-safe)
    if (["==", "!=", "<", "<=", ">", ">="].includes(operator)) {
      if (rawChildren.length !== 2) {
        diagnosticRegistry.emit({
          type: "arity_mismatch",
          nodeType: type,
          name: operator,
          expected: 2,
          actual: rawChildren.length,
          message: `Comparison operator '${operator}' expects exactly 2 operands, but got ${rawChildren.length}.`,
          node
        });
        return null;
      }
      const leftVal = evaluateAST(rawChildren[0], context, currentIndices);
      const rightVal = evaluateAST(rawChildren[1], context, currentIndices);

      // Direct equality/inequality is null-safe
      if (operator === "==") {
        const l = leftVal === undefined ? null : leftVal;
        const r = rightVal === undefined ? null : rightVal;
        return l === r;
      }
      if (operator === "!=") {
        const l = leftVal === undefined ? null : leftVal;
        const r = rightVal === undefined ? null : rightVal;
        return l !== r;
      }

      // Ordered comparison with nulls always returns false
      if (
        leftVal === null ||
        leftVal === undefined ||
        rightVal === null ||
        rightVal === undefined
      ) {
        return false;
      }

      const lNum = parseFloat(leftVal);
      const rNum = parseFloat(rightVal);

      // Fallback to string comparison if not both numeric
      const useNumeric = !isNaN(lNum) && !isNaN(rNum);
      const l = useNumeric ? lNum : String(leftVal);
      const r = useNumeric ? rNum : String(rightVal);

      if (operator === "<") return l < r;
      if (operator === "<=") return l <= r;
      if (operator === ">") return l > r;
      if (operator === ">=") return l >= r;
    }

    return null;
  }

  // 4. FUNCTION
  if (type === "FUNCTION" || type === "function") {
    const funcName = (
      typeof valAttr === "string" ? valAttr : node.operator || ""
    ).toLowerCase();

    if (funcName === "is_empty" || funcName === "empty") {
      if (rawChildren.length !== 1) {
        diagnosticRegistry.emit({
          type: "arity_mismatch",
          nodeType: type,
          name: funcName,
          expected: 1,
          actual: rawChildren.length,
          message: `Function '${funcName}' expects exactly 1 operand, but got ${rawChildren.length}.`,
          node
        });
        return null;
      }
      const childVal = evaluateAST(rawChildren[0], context, currentIndices);
      return (
        childVal === null ||
        childVal === undefined ||
        String(childVal).trim() === ""
      );
    }

    if (funcName === "is_not_empty") {
      if (rawChildren.length !== 1) {
        diagnosticRegistry.emit({
          type: "arity_mismatch",
          nodeType: type,
          name: funcName,
          expected: 1,
          actual: rawChildren.length,
          message: `Function '${funcName}' expects exactly 1 operand, but got ${rawChildren.length}.`,
          node
        });
        return null;
      }
      const childVal = evaluateAST(rawChildren[0], context, currentIndices);
      return (
        childVal !== null &&
        childVal !== undefined &&
        String(childVal).trim() !== ""
      );
    }

    if (funcName === "indexed-repeat") {
      if (rawChildren.length !== 3) {
        diagnosticRegistry.emit({
          type: "arity_mismatch",
          nodeType: type,
          name: funcName,
          expected: 3,
          actual: rawChildren.length,
          message: `Function '${funcName}' expects exactly 3 operands, but got ${rawChildren.length}.`,
          node
        });
        return null;
      }
      const targetFieldNode = rawChildren[0];
      const repeatGroupNode = rawChildren[1];
      const indexNode = rawChildren[2];

      const targetPath =
        targetFieldNode.value ||
        (targetFieldNode.field_ref && targetFieldNode.field_ref.field_id) ||
        "";
      const repeatGroup =
        repeatGroupNode.value ||
        (repeatGroupNode.field_ref && repeatGroupNode.field_ref.field_id) ||
        "";
      const indexVal = evaluateAST(indexNode, context, currentIndices);

      const targetIndex = parseInt(indexVal, 10);
      if (isNaN(targetIndex)) {
        diagnosticRegistry.emit({
          type: "type_mismatch",
          nodeType: type,
          name: funcName,
          expected: "integer",
          actual: typeof indexVal,
          message: `Function '${funcName}' expects integer index for third argument, but got non-integer value '${indexVal}'.`,
          node
        });
        return null;
      }

      const fieldName = targetPath.split("/").pop();
      const indexedPath = `${repeatGroup}[${targetIndex}]/${fieldName}`;

      return context[indexedPath] !== undefined ? context[indexedPath] : null;
    }

    // Unknown function fallback
    diagnosticRegistry.emit({
      type: "unknown_function",
      nodeType: type,
      name: funcName,
      message: `Unknown function '${funcName}'.`,
      node
    });
    return null;
  }

  return null;
}

/**
 * Compiled cached expression wrapper.
 * Returns a fast function that uses the compiled AST.
 *
 * @param {Object} node - AST ExpressionNode
 * @returns {Function} Function of form (context, currentIndices) => value
 */
export function getCompiledExpression(node) {
  if (!node) return () => null;
  const cacheKey = JSON.stringify(node);
  let compiled = compilerCache.get(cacheKey);

  if (!compiled) {
    compiled = (context, currentIndices) =>
      evaluateAST(node, context, currentIndices);
    compilerCache.put(cacheKey, compiled);
  }

  return compiled;
}

/**
 * Pure evaluation function for dynamic forms.
 * Determines field visibility, performs cascading dependent nullification,
 * and tracks field values and audit logs.
 *
 * @param {Array} fields - Array of field metadata objects
 * @param {Object} currentValues - Current state of form values
 * @param {Function} logPurge - Callback function to log field purges in the audit trail
 * @returns {Object} { visibleFields, updatedValues }
 */
export function renderFormFromJSON(fields, currentValues, logPurge = null) {
  const updatedValues = { ...currentValues };
  const visibleFields = {};

  let changed = true;
  let passes = 0;
  // Support cascading up to 10 levels deep
  while (changed && passes < 10) {
    changed = false;
    passes++;

    for (const field of fields) {
      const isRelevant = field.relevant
        ? evaluateAST(field.relevant, updatedValues) !== false
        : true;

      const wasVisible = visibleFields[field.id] !== false;
      if (visibleFields[field.id] === undefined || isRelevant !== wasVisible) {
        visibleFields[field.id] = isRelevant;
        changed = true;
      }

      if (!isRelevant) {
        const val = updatedValues[field.id];
        if (val !== undefined && val !== "" && val !== null) {
          updatedValues[field.id] = "";
          if (logPurge) {
            logPurge(
              field.id,
              val,
              "System-initiated purge of inactive child variable due to parent value mutation"
            );
          }
          changed = true;
        }
      }
    }
  }

  return { visibleFields, updatedValues };
}

/**
 * Wrapper for field validation that supports AST constraints.
 *
 * @param {Object} fieldMeta - Field metadata configuration
 * @param {any} val - Field value
 * @param {Object} context - Optional form context for constraint evaluation
 * @returns {Object} { valid: boolean, message: string }
 */
export function validateField(fieldMeta, val, context = {}) {
  return uiValidateField(fieldMeta, val, context, evaluateAST);
}
